src/components/tl/ucp/gather/gather_knomial.c

```diff
@@ -0,0 +1,171 @@
+#include "utils/ucc_math.h"
+
+#define SAVE_STATE(_phase)                                                     \
+    do {                                                                       \
+        task->gather_kn.phase = _phase;                                        \
+    } while (0)
+
+void ucc_tl_ucp_gather_knomial_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_coll_args_t *  args      = &TASK_ARGS(task);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ucc_rank_t         team_size = UCC_TL_TEAM_SIZE(team);
+    ucc_rank_t         rank      = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         size      = UCC_TL_TEAM_SIZE(team);
+    ucc_rank_t         root      = (ucc_rank_t)args->root;
+    uint32_t           radix     = task->gather_kn.radix;
+    ucc_rank_t         vrank     = (rank - root + size) % size;
+    ucc_memory_type_t  mtype     = args->src.info.mem_type;
+    size_t             data_size =
+        args->src.info.count * ucc_dt_size(args->src.info.datatype);
+    size_t     msg_size, msg_nbr;
+    void *     scratch_offset;
+    ucc_rank_t vpeer, peer, vroot_at_level, root_at_level, pos;
+    uint32_t   i;
+
+    data_size = args->src.info.count * ucc_dt_size(args->src.info.datatype);
+
+UCC_GATHER_KN_PHASE_PROGRESS:
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return;
+    }
+
+    UCC_GATHER_KN_GOTO_PHASE(task->gather_kn.phase);
+
+UCC_GATHER_KN_PHASE_INIT:
+    while (task->gather_kn.dist <= task->gather_kn.max_dist) {
+        scratch_offset = task->gather_kn.scratch;
+        if (vrank % task->gather_kn.dist == 0) {
+            pos = (vrank / task->gather_kn.dist) % radix;
+            if (pos == 0) {
+                for (i = 1; i < radix; i++) {
+                    vpeer   = vrank + i * task->gather_kn.dist;
+                    msg_nbr = ucc_min(task->gather_kn.dist, team_size - vpeer);
+                    if (vpeer >= size) {
+                        break;
+                    } else if (vrank != 0) {
+                        msg_size       = data_size * msg_nbr;
+                        scratch_offset = PTR_OFFSET(
+                            scratch_offset, data_size * task->gather_kn.dist);
+                        peer = (vpeer + root) % size;
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                         msg_size, mtype, peer,
+                                                         team, task),
+                                      task, out);
+                    } else { //The root is a particular case because it must aggregate the data sorted by ranks
+                        peer           = (vpeer + root) % size;
+                        scratch_offset = PTR_OFFSET(task->gather_kn.scratch,
+                                                    data_size * peer);
+                        if (msg_nbr <=
```

Comment:
keep single line: if (msg_nbr <= team_size - peer) {

(comment can be on a separate line before)